heron/api/src/java/com/twitter/heron/api/bolt/WindowedBoltExecutor.java

```diff
@@ -0,0 +1,415 @@
+   */
+  protected void start() {
+    if (waterMarkEventGenerator != null) {
+      LOG.fine("Starting waterMarkEventGenerator");
+      waterMarkEventGenerator.start();
+    }
+    LOG.fine("Starting trigger policy");
+    triggerPolicy.start();
+  }
+
+  private boolean isTupleTs() {
+    return timestampExtractor != null;
+  }
+
+  @SuppressWarnings("HiddenField")
+  private TriggerPolicy<Tuple, ?> getTriggerPolicy(Count slidingIntervalCount, Long
+      slidingIntervalDurationMs, WindowManager<Tuple> manager, EvictionPolicy<Tuple, ?>
+      evictionPolicy) {
+    if (slidingIntervalCount != null) {
+      if (isTupleTs()) {
+        return new WatermarkCountTriggerPolicy<>(slidingIntervalCount.value, manager,
+            evictionPolicy, manager);
+      } else {
+        return new CountTriggerPolicy<>(slidingIntervalCount.value, manager, evictionPolicy);
+      }
+    } else {
+      if (isTupleTs()) {
+        return new WatermarkTimeTriggerPolicy<>(slidingIntervalDurationMs, manager,
+            evictionPolicy, manager);
+      } else {
+        return new TimeTriggerPolicy<>(slidingIntervalDurationMs, manager, evictionPolicy);
+      }
+    }
+  }
+
+  @SuppressWarnings("HiddenField")
+  private EvictionPolicy<Tuple, ?> getEvictionPolicy(Count windowLengthCount, Long
+      windowLengthDurationMs) {
+    if (windowLengthCount != null) {
+      if (isTupleTs()) {
+        return new WatermarkCountEvictionPolicy<>(windowLengthCount.value);
+      } else {
+        return new CountEvictionPolicy<>(windowLengthCount.value);
+      }
+    } else {
+      if (isTupleTs()) {
+        return new WatermarkTimeEvictionPolicy<>(windowLengthDurationMs, maxLagMs);
+      } else {
+        return new TimeEvictionPolicy<>(windowLengthDurationMs);
+      }
+    }
+  }
+
+  @Override
+  public void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector
+      collector) {
+    doPrepare(topoConf, context, collector, new ConcurrentLinkedQueue<>());
+  }
+
+  // NOTE: the queue has to be thread safe.
+  protected void doPrepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector
+      collector, Collection<Event<Tuple>> queue) {
+    Objects.requireNonNull(topoConf);
+    Objects.requireNonNull(context);
+    Objects.requireNonNull(collector);
+    Objects.requireNonNull(queue);
+    this.windowedOutputCollector = new WindowedOutputCollector(collector);
+    bolt.prepare(topoConf, context, windowedOutputCollector);
+    this.listener = newWindowLifecycleListener();
+    this.windowManager = initWindowManager(listener, topoConf, context, queue);
+    start();
+    LOG.info(String.format("Initialized window manager %s", windowManager));
+  }
+
+  @Override
+  public void execute(Tuple input) {
+    if (isTupleTs()) {
+      long ts = timestampExtractor.extractTimestamp(input);
+      if (waterMarkEventGenerator.track(input.getSourceGlobalStreamId(), ts)) {
```

Comment:
Could it be possible not to use getSourceGlobalStreamId(). One of the reasons that we did not implement this in heron was that we felt it exposed too low level details. Could we instead get the SourceComponent and SourceStreamId as individual parameters?

Reply:
@srkukarni I understand what you are saying, but there are already places in the Heron Java API that use proto buf objects e.g. TopologyContext:

Map<TopologyAPI.StreamId, TopologyAPI.Grouping> getThisSources()